feat: its scripts (#12)
eigerco/axelar-contract-deployments: starknet/docs/its.md

  -y, --yes                          skip deployment prompt confirmation (env:
                                     YES)
  -p, --privateKey < privateKey >    private key for Starknet account(testnet
                                     only, not required for offline tx
                                     generation) (env: STARKNET_PRIVATE_KEY)
  --accountAddress <accountAddress>  Starknet account address (env:
                                     STARKNET_ACCOUNT_ADDRESS)
  -h, --help                         display help for command
```

**Note:** Unlike other deployment scripts, ITS deployment does not support offline mode or gas estimation as it uses direct contract deployment.

**Example:**
```bash
npx ts-node starknet/its/deploy-its.ts \
  --env testnet \
  --gateway 0x... \
  --gasService 0x... \
  --chainName starknet \
  --itsHubChainName axelar \
  --itsHubContractAddress 0x... \
  --operator 0x... \
  --owner 0x... \
  --salt 0x123 \
  --privateKey 0x... \
  --accountAddress 0x...
```

### Deploy InterchainTokenFactory

Deploy the InterchainTokenFactory contract on Starknet. This contract is responsible for creating and managing interchain tokens.

```

> Usage: deploy-itf [options]

Reviewer: Ah! makes sense

                                     generation) (env: STARKNET_PRIVATE_KEY)
  --accountAddress <accountAddress>  Starknet account address (env:
                                     STARKNET_ACCOUNT_ADDRESS)
  -h, --help                         display help for command
```

**Note:** Like ITS deployment, ITF deployment does not support offline mode or gas estimation as it uses direct contract deployment.

**Example:**
```bash
npx ts-node starknet/its/deploy-itf.ts \
  --env testnet \
  --interchainTokenService 0x... \
  --owner 0x... \
  --salt 0x123 \
  --privateKey 0x... \
  --accountAddress 0x...
```

## Token Operations

### Deploy Interchain Token

Deploy a new interchain token locally via the InterchainTokenFactory.

```
Usage: deploy-token [options]

Deploy a new interchain token on Starknet via InterchainTokenFactory

Options:
  --salt <salt>                      Salt for token deployment
  --name <name>                      Token name
  --symbol <symbol>                  Token symbol
  --decimals <decimals>              Token decimals
  --initialSupply <amount>           Initial supply to mint (defaults to 0)

>   --minter <address>                 Minter address (if initial supply > 0,
>                                      mintership will be transferred to this
>                                      address)

Reviewer: ```suggestion
  --minter <address>                 Minter address (mintership will be
                                     transferred to this address)
```

I believe whenever a minter is passed, the mintership is transferred. It reverts in case the initial supply is <= 0 and no minter is passed.

                                     "testnet", env: ENV)
  -y, --yes                          skip deployment prompt confirmation (env:
                                     YES)
  -p, --privateKey < privateKey >    private key for Starknet account(testnet
                                     only, not required for offline tx
                                     generation) (env: STARKNET_PRIVATE_KEY)
  --accountAddress <accountAddress>  Starknet account address (env:
                                     STARKNET_ACCOUNT_ADDRESS)
  -h, --help                         display help for command

Note: This command deploys tokens locally via the InterchainTokenFactory.
- If initialSupply > 0, tokens will be minted to the deployer and mintership transferred to the minter
- If initialSupply = 0, the minter address will be set directly
- For cross-chain deployment, use a separate command after local deployment
```

**Examples:**

Deploy with initial supply:
```bash
npx ts-node starknet/its/deploy-token.ts --salt my-token-salt --name "My Token" --symbol "MTK" --decimals 18 --initialSupply 1000000 --minter 0x123... --privateKey 0x... --accountAddress 0x...
```

> Deploy without initial supply (minter gets set directly):

Reviewer: ```suggestion
Deploy without initial supply (minter is required in this case):
```
transferred or set directly is an implementation detail, I think we don't need to mention it here. What matters is that if the initial supply is not set, a minter has to be passed otherwise it'll revert.

Author: I also added [a check](https://github.com/eigerco/axelar-contract-deployments/pull/12/commits/85e4aa2a2271a65699f4a0160003ba62e21208c9) for it

  -e, --env <env>                    environment (choices: "devnet-amplifier",
                                     "mainnet", "stagenet", "testnet", default:
                                     "testnet", env: ENV)
  -y, --yes                          skip deployment prompt confirmation (env:
                                     YES)
  -p, --privateKey < privateKey >    private key for Starknet account(testnet
                                     only, not required for offline tx
                                     generation) (env: STARKNET_PRIVATE_KEY)
  --accountAddress <accountAddress>  Starknet account address (env:
                                     STARKNET_ACCOUNT_ADDRESS)
  -h, --help                         display help for command
```

**Example:**
```bash
npx ts-node starknet/its/register-token.ts --salt my-salt --tokenAddress 0x... --tokenManagerType lockUnlock --privateKey 0x... --accountAddress 0x...
```

### Register Canonical Token

```
Usage: register-canonical-token [options]

Register a canonical token for cross-chain use (uses lock/unlock mechanism)

Options:
  --tokenAddress <address>           Address of the canonical token to register

>   --factoryAddress <address>         InterchainTokenFactory address (defaults

Reviewer: Why is this here but not in the other commands? :thinking:

Author: Factory address is passed for `deploy-remote-canonical-token`, `register-canonical-token`, `manage-service` set-factory-address action and `calculate-token-id`

                                     "testnet", env: ENV)
  -y, --yes                          skip deployment prompt confirmation (env:
                                     YES)
  -p, --privateKey < privateKey >    private key for Starknet account(testnet
                                     only, not required for offline tx
                                     generation) (env: STARKNET_PRIVATE_KEY)
  --accountAddress <accountAddress>  Starknet account address (env:
                                     STARKNET_ACCOUNT_ADDRESS)
  -h, --help                         display help for command
```

**Example:**
```bash
npx ts-node starknet/its/register-canonical-token.ts --tokenAddress 0x... --privateKey 0x... --accountAddress 0x...
```

## Cross-Chain Operations

### Interchain Transfer

```
Usage: transfer [options]

Transfer tokens across chains using InterchainTokenService

Options:
  --tokenId <id>                     Token ID (hex string)
  --destinationChain <chain>         Destination chain name
  --destinationAddress <address>     Destination address
  --amount <amount>                  Amount to transfer (in smallest unit)

>   --data <data>                      Optional data for contract execution

Author: I actually made data to be a hex string representing the bytes they want to pass. 
TLDR - we take the hex string, convert to bytes, convert them to string and then `byteArray.byteArrayFromString(...)` that string
